File: service-render/src/render/numbering.py
"""Document numbering (FR-EDT-07/08, FR-EXP-02): the same counter rules as
the client's `features/documents/editor/numbering.ts`, ported here so export
produces exactly what the editor showed — one spec, two implementations,
per FR-EXP-02.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

# ДСТУ 3008:2015 appendix lettering — Ukrainian alphabet excluding
# Ґ, Є, З, І, Ї, Й, О, Ч, Ь. Must match numbering.ts APPENDIX_LETTERS exactly.
APPENDIX_LETTERS = [
    "А", "Б", "В", "Г", "Д", "Е", "Ж", "И", "К", "Л", "М", "Н", "П", "Р", "С", "Т",
    "У", "Ф", "Х", "Ц", "Ш", "Щ", "Ю", "Я",
]

SectionKind = Literal["chapter", "appendix"]


# What a numbered block *is*, so a cross-reference can pick its wording
# («рис. 2.1» vs «табл. 2.1») without the reference node storing it.
NumberedKind = Literal["heading", "figure", "table", "formula"]


@dataclass
class NumberingInput:
    id: str
    kind: SectionKind
    order: int
    content: dict[str, Any] | None


@dataclass
class NumberingResult:
    section_labels: dict[str, str]
    block_numbers: dict[str, str]
    section_kinds: dict[str, SectionKind]
    block_kinds: dict[str, NumberedKind]


def _appendix_label(index: int) -> str:
    if index < len(APPENDIX_LETTERS):
        return APPENDIX_LETTERS[index]
    return f"Дод.{index + 1}"


def _walk(node: dict[str, Any], fn) -> None:
    fn(node)
    for child in node.get("content") or []:
        _walk(child, fn)


def compute_numbering(sections: list[NumberingInput]) -> NumberingResult:
    section_labels: dict[str, str] = {}
    block_numbers: dict[str, str] = {}
    section_kinds: dict[str, SectionKind] = {}
    block_kinds: dict[str, NumberedKind] = {}

    def record(block_id: str, number: str, kind: NumberedKind) -> None:
        block_numbers[block_id] = number
        block_kinds[block_id] = kind

    chapter_index = 0
    appendix_index = 0

    for section in sorted(sections, key=lambda s: s.order):
        if section.kind == "appendix":
            label = _appendix_label(appendix_index)
            appendix_index += 1
        else:
            chapter_index += 1
            label = str(chapter_index)
        section_labels[section.id] = label
        section_kinds[section.id] = section.kind

        if not section.content:
            continue

        heading_counters = [0, 0, 0]  # levels 2, 3, 4
        figure_counter = 0
        table_counter = 0
        formula_counter = 0

        def visit(node: dict[str, Any]) -> None:
            nonlocal figure_counter, table_counter, formula_counter
            block_id = (node.get("attrs") or {}).get("blockId")
            if not block_id:
                return

            node_type = node.get("type")
            if node_type == "heading":
                level = int((node.get("attrs") or {}).get("level", 1))
                if level < 2 or level > 4:
                    return
                idx = level - 2
                heading_counters[idx] += 1
                for j in range(idx + 1, len(heading_counters)):
                    heading_counters[j] = 0
                record(block_id, ".".join([label, *[str(c) for c in heading_counters[: idx + 1]]]), "heading")
                return

            if node_type == "image":
                figure_counter += 1
                record(block_id, f"{label}.{figure_counter}", "figure")
                return

            if node_type == "table":
                table_counter += 1
                record(block_id, f"{label}.{table_counter}", "table")
                return

            if node_type == "formulaBlock":
                formula_counter += 1
                record(block_id, f"{label}.{formula_counter}", "formula")

        _walk(section.content, visit)

    return NumberingResult(
        section_labels=section_labels,
        block_numbers=block_numbers,
        section_kinds=section_kinds,
        block_kinds=block_kinds,
    )


def figure_caption(number: str, caption: str) -> str:
    return f"Рисунок {number} — {caption}" if caption else f"Рисунок {number}"


def table_caption(number: str, caption: str) -> str:
    return f"Таблиця {number} — {caption}" if caption else f"Таблиця {number}"


def table_continuation_caption(number: str) -> str:
    """Heading above a table that continues on a following page (FR-TBL-09)."""
    return f"Продовження таблиці {number}"


def formula_number(number: str) -> str:
    """Formula numbers are bare parenthesised counters, right-aligned on the
    formula's line (ДСТУ 3008:2015)."""
    return f"({number})"


def section_label(kind: SectionKind, label: str) -> str:
    return f"Додаток {label}" if kind == "appendix" else f"Розділ {label}"


# In-text cross-reference wording (FR-EDT-04/07). Short forms, per Ukrainian
# academic practice: a caption reads «Рисунок 2.1 — …», a reference to it reads
# «рис. 2.1». Must match numbering.ts `referenceText` exactly.
_REFERENCE_PREFIX: dict[str, str] = {
    "heading": "розд. ",
    "figure": "рис. ",
    "table": "табл. ",
}

# Shown when the referenced block was deleted; export validation blocks on it,
# so it should never reach a finished document.
UNRESOLVED_REFERENCE = "[?]"


def reference_labels(result: NumberingResult) -> dict[str, str]:
    """targetId → rendered label, for every block and section a cross-reference
    can point at. Mirrors `referenceLabels()` in the client's numbering.ts."""
    labels: dict[str, str] = {}

    for block_id, number in result.block_numbers.items():
        kind = result.block_kinds.get(block_id)
        if kind == "formula":
            labels[block_id] = f"({number})"
        elif kind in _REFERENCE_PREFIX:
            labels[block_id] = f"{_REFERENCE_PREFIX[kind]}{number}"

    for section_id, label in result.section_labels.items():
        kind = result.section_kinds.get(section_id, "chapter")
        labels[section_id] = f"додаток {label}" if kind == "appendix" else f"розділ {label}"

    return labels

File: client/src/app/features/documents/editor/numbering.ts
/**
 * Live document numbering (FR-EDT-07/08): headings, figures, and tables number
 * from document structure, never hand-typed. Computed here (not stored) from
 * the ordered list of sections' ProseMirror JSON, shared by the editor
 * decoration layer and (eventually) the export translator's fixture suite
 * (FR-EXP-02 calls for one counter spec, two implementations).
 */

// ДСТУ 3008:2015 appendix lettering — Ukrainian alphabet excluding
// Ґ, Є, З, І, Ї, Й, О, Ч, Ь (letters easily confused with digits/Latin or
// without a distinct print form in this context).
export const APPENDIX_LETTERS = [
  'А',
  'Б',
  'В',
  'Г',
  'Д',
  'Е',
  'Ж',
  'И',
  'К',
  'Л',
  'М',
  'Н',
  'П',
  'Р',
  'С',
  'Т',
  'У',
  'Ф',
  'Х',
  'Ц',
  'Ш',
  'Щ',
  'Ю',
  'Я',
];

export type SectionKindForNumbering = 'chapter' | 'appendix';

export interface NumberingInput {
  id: string;
  kind: SectionKindForNumbering;
  order: number;
  content: PMNode | null;
  /** Only used to describe the section in the cross-reference picker. */
  title?: string;
}

/** Minimal shape of a ProseMirror JSON node, enough to walk for numbering. */
export interface PMNode {
  type?: string;
  text?: string;
  attrs?: Record<string, unknown>;
  content?: PMNode[];
}

/** What a numbered block *is*, so a cross-reference can pick its wording
 * («рис. 2.1» vs «табл. 2.1») without the reference node storing it — a stored
 * kind would go stale the moment the target changed. */
export type NumberedKind = 'heading' | 'figure' | 'table' | 'formula';

export interface NumberingResult {
  /** section id -> chapter number ("2") or appendix letter ("А") */
  sectionLabels: Map<string, string>;
  /** section id -> chapter | appendix, for section-level reference wording */
  sectionKinds: Map<string, SectionKindForNumbering>;
  /** block_id -> computed label ("2.1.3" for a heading, "2.1" for a figure/table) */
  blockNumbers: Map<string, string>;
  /** block_id -> what that block is (FR-EDT-04 cross-references) */
  blockKinds: Map<string, NumberedKind>;
}

function appendixLabel(index: number): string {
  return APPENDIX_LETTERS[index] ?? `Дод.${index + 1}`;
}

function walk(node: PMNode, fn: (node: PMNode) => void): void {
  fn(node);
  for (const child of node.content ?? []) walk(child, fn);
}

export function computeNumbering(sections: NumberingInput[]): NumberingResult {
  const sectionLabels = new Map<string, string>();
  const sectionKinds = new Map<string, SectionKindForNumbering>();
  const blockNumbers = new Map<string, string>();
  const blockKinds = new Map<string, NumberedKind>();

  const record = (blockId: string, number: string, kind: NumberedKind): void => {
    blockNumbers.set(blockId, number);
    blockKinds.set(blockId, kind);
  };

  let chapterIndex = 0;
  let appendixIndex = 0;

  const ordered = [...sections].sort((a, b) => a.order - b.order);

  for (const section of ordered) {
    const label =
      section.kind === 'appendix' ? appendixLabel(appendixIndex++) : String(++chapterIndex);
    sectionLabels.set(section.id, label);
    sectionKinds.set(section.id, section.kind);

    if (!section.content) continue;

    const headingCounters = [0, 0, 0]; // heading levels 2, 3, 4
    let figureCounter = 0;
    let tableCounter = 0;
    let formulaCounter = 0;

    walk(section.content, (node) => {
      const blockId = node.attrs?.['blockId'] as string | undefined;
      if (!blockId) return;

      if (node.type === 'heading') {
        const level = Number(node.attrs?.['level'] ?? 1);
        if (level < 2 || level > 4) return;
        const idx = level - 2;
        headingCounters[idx]++;
        for (let j = idx + 1; j < headingCounters.length; j++) headingCounters[j] = 0;
        record(blockId, [label, ...headingCounters.slice(0, idx + 1)].join('.'), 'heading');
        return;
      }

      if (node.type === 'image') {
        figureCounter++;
        record(blockId, `${label}.${figureCounter}`, 'figure');
        return;
      }

      if (node.type === 'table') {
        tableCounter++;
        record(blockId, `${label}.${tableCounter}`, 'table');
        return;
      }

      if (node.type === 'formulaBlock') {
        formulaCounter++;
        record(blockId, `${label}.${formulaCounter}`, 'formula');
      }
    });
  }

  return { sectionLabels, sectionKinds, blockNumbers, blockKinds };
}

/** Caption text conventions (FR-EDT-08) — centralized so editor rendering and
 * (later) docx export use identical wording. */
export const captionText = {
  figure: (number: string, caption: string): string =>
    caption ? `Рисунок ${number} — ${caption}` : `Рисунок ${number}`,
  table: (number: string, caption: string): string =>
    caption ? `Таблиця ${number} — ${caption}` : `Таблиця ${number}`,
  /** Formula numbers are bare parenthesised counters, right-aligned on the
   * formula's line (ДСТУ 3008:2015). */
  formula: (number: string): string => `(${number})`,
  sectionLabel: (kind: SectionKindForNumbering, label: string): string =>
    kind === 'appendix' ? `Додаток ${label}` : `Розділ ${label}`,
};

/** In-text cross-reference wording (FR-EDT-04/07). Short forms, per Ukrainian
 * academic practice: a caption reads «Рисунок 2.1 — …», a reference to it
 * reads «рис. 2.1». */
export const referenceText = {
  heading: (number: string): string => `розд. ${number}`,
  figure: (number: string): string => `рис. ${number}`,
  table: (number: string): string => `табл. ${number}`,
  formula: (number: string): string => `(${number})`,
  section: (kind: SectionKindForNumbering, label: string): string =>
    kind === 'appendix' ? `додаток ${label}` : `розділ ${label}`,
  /** Shown when the referenced block was deleted — the same orphan signal
   * citations use, and it blocks export. */
  unresolved: '[?]',
};

/**
 * Rendered label for every referenceable target, keyed by the id a
 * cross-reference node stores. Cross-references keep only that id: the wording
 * and the number are both derived, so moving or renumbering a figure updates
 * every reference to it without touching stored content (FR-EDT-07).
 */
export function referenceLabels(result: NumberingResult): Map<string, string> {
  const labels = new Map<string, string>();

  for (const [blockId, number] of result.blockNumbers) {
    const kind = result.blockKinds.get(blockId);
    if (kind) labels.set(blockId, referenceText[kind](number));
  }
  for (const [sectionId, label] of result.sectionLabels) {
    labels.set(
      sectionId,
      referenceText.section(result.sectionKinds.get(sectionId) ?? 'chapter', label)
    );
  }

  return labels;
}

/** One entry in the "insert cross-reference" picker. */
export interface ReferenceTarget {
  /** block id, or section id for a whole section */
  id: string;
  kind: NumberedKind | 'section';
  /** what the reference will render as, e.g. «рис. 2.1» */
  label: string;
  /** caption / heading text, so the picker is readable */
  description: string;
}

function inlineText(node: PMNode): string {
  return (node.text ?? '') + (node.content ?? []).map(inlineText).join('');
}

/** Everything a student can point a cross-reference at, in document order. */
export function collectReferenceTargets(
  sections: NumberingInput[],
  result: NumberingResult
): ReferenceTarget[] {
  const targets: ReferenceTarget[] = [];
  const ordered = [...sections].sort((a, b) => a.order - b.order);

  for (const section of ordered) {
    const sectionLabel = result.sectionLabels.get(section.id);
    if (sectionLabel) {
      targets.push({
        id: section.id,
        kind: 'section',
        label: referenceText.section(section.kind, sectionLabel),
        description: section.title ?? '',
      });
    }
    if (!section.content) continue;

    walk(section.content, (node) => {
      const blockId = node.attrs?.['blockId'] as string | undefined;
      if (!blockId) return;
      const kind = result.blockKinds.get(blockId);
      const number = result.blockNumbers.get(blockId);
      if (!kind || !number) return;

      const caption = (node.attrs?.['caption'] as string) ?? '';
      const description =
        kind === 'heading'
          ? inlineText(node)
          : kind === 'formula'
            ? ((node.attrs?.['latex'] as string) ?? '')
            : caption;

      targets.push({ id: blockId, kind, label: referenceText[kind](number), description });
    });
  }

  return targets;
}
